docs/generated/UNDERLAY_CONFIG.md

```diff
@@ -188,6 +193,136 @@ Name of the field or column name that maps to the occurrence entity id. Required
 
 
 
+## SZCriteriaSelector
+Criteria selector configuration.
+
+Define a version of this file for each set of UI plugins + configuration.
+
+### SZCriteriaSelector.display
+**required** [SZCriteriaSelectorDisplay](#szcriteriaselectordisplay)
+
+Display information.
+
+### SZCriteriaSelector.filterBuilder
+**required** String
+
+Name of a Java class that implements the `FilterBuilder` interface. This class will take in the selector configuration and user selections and produce an `EntityFilter` on either the primary entity (for a cohort) or another entity (for a data feature).
+
+### SZCriteriaSelector.isEnabledForCohorts
+**required** boolean
+
+True if this criteria selector should be displayed in the cohort builder.
+
+### SZCriteriaSelector.isEnabledForDataFeatureSets
+**required** boolean
+
+True if this criteria selector should be displayed in the data feature set builder.
+
+### SZCriteriaSelector.modifiers
+**required** List [ SZCriteriaSelector$Modifier ]
+
+Configuration for modifiers.
+
+### SZCriteriaSelector.name
+**required** String
+
+Name of the criteria selector.
+
+This is the unique identifier for the selector. The selector names cannot overlap within an underlay.
+
+Name may not include spaces or special characters, only letters and numbers.
+
+This name is stored in the application database for cohorts and data feature sets, so once there are artifacts associated with a criteria selector, you can't change the selector name.
+
+### SZCriteriaSelector.plugin
+**required** String
+
+Name of the primary UI display plugin. (e.g. selector for condition, not any of the modifiers).
+
+This plugin name is stored in the application database, so once there are cohorts or data features that use this selector, you can't change the plugin names.
+
+### SZCriteriaSelector.pluginConfig
+**required** String
+
+Serialized configuration of the primary UI display plugin e.g. "{"attribute":"gender"}".
```

Comment:
Might be worth changing things like this to reference serialized protos instead of JSON.

Reply:
Are you talking about the example JSON, or the term "serialized configuration", or both?

I was expecting to continue using JSON as the serialization format, to be consistent with the rest of the underlay config, and then deserialize using a protobuf utility (in Java, [JsonFormat](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/util/JsonFormat)).

Reply:
Going to merge this pr now, and make any changes in a follow-on pr once we finish discussing.

Reply:
For the config files, I lean towards JSON over another human readable format, because that's what the other config files use.

For the `/v2/underlay/{underlayName}` API, the plugin configs are currently also being returned as JSON. I could change that, now that I'll have the schemas -- I don't think the file & wire formats need to match. Although given we're not using protos for the underlay config (yet anyway), we'd be returning an object with properties that have been serialized differently (e.g. entities in JSON, plugin config in proto packed).

For the storage in postgres, the plugin data is currently being stored as it comes over the wire. I could change this too though, now that I'll have the schemas -- I don't think the wire & storage formats need to match. Though I'm not expecting these plugin configs to be huge, so maybe better to prioritize easier debugging at this point.

Reply:
> For the config files, I lean towards JSON over another human readable format, because that's what the other config files use.

I agree, JSON makes sense.

> 
> For the `/v2/underlay/{underlayName}` API, the plugin configs are currently also being returned as JSON. I could change that, now that I'll have the schemas -- I don't think the file & wire formats need to match. Although given we're not using protos for the underlay config (yet anyway), we'd be returning an object with properties that have been serialized differently (e.g. entities in JSON, plugin config in proto packed).

To make sure we're on the same page, the selector information with the name etc. is JSON that comes through the SZxxx classes and I agree should definitely stay the way it is. Within that there's the plugin specific proto config which will be serialized as a string, either JSON or packed proto. Are you suggesting we should serialize that as JSON as well so they're the same, even though it'll be serialized into a string? I don't think I feel strongly either way so we could go with JSON everywhere.

> 
> For the storage in postgres, the plugin data is currently being stored as it comes over the wire. I could change this too though, now that I'll have the schemas -- I don't think the wire & storage formats need to match. Though I'm not expecting these plugin configs to be huge, so maybe better to prioritize easier debugging at this point.

I could switch the selection data to store as JSON in that case, but I'll maybe leave that to the last step and drop backwards compatibility with the non-proto data at the same time.